pkg/datasources/grants.go

```diff
@@ -248,135 +321,193 @@ var grantsSchema = map[string]*schema.Schema{
+		// if inApplicationPackage := shareMap["in_application_package"]; inApplicationPackage != "" {
+		//	opts.To.Share.InApplicationPackage = sdk.Pointer(sdk.NewAccountObjectIdentifier(inApplicationPackage.(string)))
+		// }
+	}
+	return opts, nil
+}
+
+func buildOptsForGrantsOf(grantsOf map[string]interface{}) (*sdk.ShowGrantOptions, error) {
+	opts := new(sdk.ShowGrantOptions)
 
-			grantDetails, err = snowflake.ShowFutureGrantsIn(db, "SCHEMA", schemaName)
-			if err != nil {
-				return err
-			}
+	if role := grantsOf["role"].(string); role != "" {
+		opts.Of = &sdk.ShowGrantsOf{
+			Role: sdk.NewAccountObjectIdentifier(role),
+		}
+	}
+	if databaseRole := grantsOf["database_role"].(string); databaseRole != "" {
+		opts.Of = &sdk.ShowGrantsOf{
+			DatabaseRole: sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(databaseRole),
+		}
+	}
+	if applicationRole := grantsOf["application_role"].(string); applicationRole != "" {
+		opts.Of = &sdk.ShowGrantsOf{
+			ApplicationRole: sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(applicationRole),
+		}
+	}
+	if share := grantsOf["share"].(string); share != "" {
+		opts.Of = &sdk.ShowGrantsOf{
+			Share: sdk.NewAccountObjectIdentifier(share),
 		}
 	}
+	return opts, nil
+}
 
-	if v, ok := d.GetOk("future_grants_to"); ok {
-		futureGrantsTo := v.([]interface{})[0].(map[string]interface{})
-		role := futureGrantsTo["role"].(string)
-		if role != "" {
-			grantDetails, err = snowflake.ShowFutureGrantsTo(db, "ROLE", role)
-			if err != nil {
-				return err
-			}
+func buildOptsForFutureGrantsIn(futureGrantsIn map[string]interface{}) (*sdk.ShowGrantOptions, error) {
+	opts := new(sdk.ShowGrantOptions)
+	opts.Future = sdk.Bool(true)
+
+	if db := futureGrantsIn["database"].(string); db != "" {
+		opts.In = &sdk.ShowGrantsIn{
+			Database: sdk.Pointer(sdk.NewAccountObjectIdentifier(db)),
+		}
+	}
+	if sc := futureGrantsIn["schema"].(string); sc != "" {
+		opts.In = &sdk.ShowGrantsIn{
+			Schema: sdk.Pointer(sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(sc)),
 		}
 	}
+	return opts, nil
+}
 
-	err = d.Set("grants", flattenGrants(grantDetails))
-	if err != nil {
-		return err
+func buildOptsForFutureGrantsTo(futureGrantsTo map[string]interface{}) (*sdk.ShowGrantOptions, error) {
+	opts := new(sdk.ShowGrantOptions)
+	opts.Future = sdk.Bool(true)
+
+	if role := futureGrantsTo["role"].(string); role != "" {
+		opts.To = &sdk.ShowGrantsTo{
+			Role: sdk.NewAccountObjectIdentifier(role),
+		}
 	}
-	d.SetId("grants")
-	return nil
+	if databaseRole := futureGrantsTo["database_role"].(string); databaseRole != "" {
+		opts.To = &sdk.ShowGrantsTo{
+			DatabaseRole: sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(databaseRole),
+		}
+	}
+	return opts, nil
 }
 
-func flattenGrants(grants []snowflake.GrantDetail) []map[string]interface{} {
+func flattenGrants(grants []sdk.Grant) []map[string]interface{} {
```

Comment:
flattening is more reminding me of flattening the structure (taking the element's children and putting everything next to the parent in one array). Maybe something like `convertGrantsToMap` or inlining the function would be better?